### PR TITLE
Introducing internal commands for SyGuS commands

### PR DESCRIPTION
**Pull request by HanielB**
This commit introduces CVC4 commands corresponding to commands in SyGuS scripts. It also changes the parser to use these commands instead of the ad-hoc code from before. The commands are handled in the SMT engine as they are invoked by CVC4's main module.

This commit will allow the future integration of SyGuS to CVC4's public API, as well as the introduction of a SyGuS printer.

### CODE REVIEW

File: src/smt/command.h [hunk 1]
```diff
@@ -613,14 +613,140 @@ class CVC4_PUBLIC QueryCommand : public Command
...
+  std::string getCommandName() const override;
+}; /* class DeclareFunctionCommand */
+
+class CVC4_PUBLIC SynthFunCommand : public DeclarationDefinitionCommand
+{
+ protected:
+  Expr d_func;
+  /** sygus type of the function-to-synthesize
+   *
+   * Note that
```
Comment (ajreynol) [inline src/smt/command.h]: Missing comment

File: src/smt/command.h [hunk 2]
```diff
@@ -613,14 +613,140 @@ class CVC4_PUBLIC QueryCommand : public Command
   std::string getCommandName() const override;
 }; /* class QueryCommand */
 
-class CVC4_PUBLIC CheckSynthCommand : public Command
+/* ------------------- sygus commands  ------------------ */
+
+class CVC4_PUBLIC DeclareVarCommand : public DeclarationDefinitionCommand
```
Comment (ajreynol) [inline src/smt/command.h]: Might want to call this DeclareSygusVarCommand to avoid confusion.

File: src/smt/command.h [hunk 3]
```diff
@@ -613,14 +613,140 @@ class CVC4_PUBLIC QueryCommand : public Command
...
+
+  /** default interface */
+  void invoke(SmtEngine* smtEngine) override;
+  Command* exportTo(ExprManager* exprManager,
+                    ExprManagerMapCollection& variableMap) override;
+  Command* clone() const override;
+  std::string getCommandName() const override;
+};
+
+class CVC4_PUBLIC DeclarePrimedVarCommand : public DeclarationDefinitionCommand
```
Comment (ajreynol) [inline src/smt/command.h]: Simliarly, DeclareSygusPrimedVarCommand

File: src/smt/command.h [hunk 4]
```diff
@@ -613,14 +613,140 @@ class CVC4_PUBLIC QueryCommand : public Command
...
+
+  /** default interface */
+  void invoke(SmtEngine* smtEngine) override;
+  Command* exportTo(ExprManager* exprManager,
+                    ExprManagerMapCollection& variableMap) override;
+  Command* clone() const override;
+  std::string getCommandName() const override;
+}; /* class DeclareFunctionCommand */
+
+class CVC4_PUBLIC ConstraintCommand : public Command
```
Comment (ajreynol) [inline src/smt/command.h]: SygusConstraintCommand? (not sure)

File: src/smt/command.h [hunk 5]
```diff
@@ -613,14 +613,140 @@ class CVC4_PUBLIC QueryCommand : public Command
...
+
+  /** default interface */
+  void invoke(SmtEngine* smtEngine) override;
+  Command* exportTo(ExprManager* exprManager,
+                    ExprManagerMapCollection& variableMap) override;
+  Command* clone() const override;
+  std::string getCommandName() const override;
+};
+
+class CVC4_PUBLIC InvConstraintCommand : public Command
```
Comment (ajreynol) [inline src/smt/command.h]: SygusInvConstraintCommand?

File: src/smt/smt_engine.cpp [hunk 6]
```diff
@@ -3706,6 +3723,232 @@ Result SmtEngine::assertFormula(const Expr& ex, bool inUnsatCore)
...
+#endif
+  Trace("smt") << "SmtEngine::declareSygusPrimedVar: " << id << "\n";
+}
+
+void SmtEngine::declareSygusFunctionVar(const std::string& id,
+                                        Expr var,
+                                        Type type)
+{
+  d_private->d_sygusVars.push_back(Node::fromExpr(var));
+  Trace("smt") << "SmtEngine::declareSygusVar: " << var << "\n";
```
Comment (ajreynol) [inline src/smt/smt_engine.cpp]: FunctionVar?

File: src/smt/smt_engine.cpp [hunk 7]
```diff
@@ -3706,6 +3723,232 @@ Result SmtEngine::assertFormula(const Expr& ex, bool inUnsatCore)
...
+    var_nodes.push_back(Node::fromExpr(v));
+  }
+  d_private->d_sygusFunVars[fn] = var_nodes;
+  // whether sygus type encodes syntax restrictions
+  if (sygusType.isDatatype()
+      && static_cast<DatatypeType>(sygusType).getDatatype().isSygus())
+  {
+    d_private->d_sygusFunSyntax[fn] = TypeNode::fromType(sygusType);
+  }
+  Trace("smt") << "SmtEngine::declareSythFun: " << func << "\n";
```
Comment (ajreynol) [inline src/smt/smt_engine.cpp]: SynthFun

File: src/smt/smt_engine.cpp [hunk 8]
```diff
@@ -3706,6 +3723,232 @@ Result SmtEngine::assertFormula(const Expr& ex, bool inUnsatCore)
...
+  d_private->d_sygusPrimedVarTypes.push_back(type);
+#endif
+  Trace("smt") << "SmtEngine::declareSygusPrimedVar: " << id << "\n";
+}
+
+void SmtEngine::declareSygusFunctionVar(const std::string& id,
+                                        Expr var,
+                                        Type type)
+{
+  d_private->d_sygusVars.push_back(Node::fromExpr(var));
```
Comment (ajreynol) [inline src/smt/smt_engine.cpp]: I'm wondering why this is the same vector as in declareSygusVar, does this preserve the previous behavior?

Comment (HanielB) [reply src/smt/smt_engine.cpp]: Yes, declared functions should be universally quantified.

File: src/smt/command.h [hunk 9]
```diff
@@ -613,29 +613,250 @@ class CVC4_PUBLIC QueryCommand : public Command
   std::string getCommandName() const override;
 }; /* class QueryCommand */
 
-class CVC4_PUBLIC CheckSynthCommand : public Command
+/* ------------------- sygus commands  ------------------ */
+
+/** Declares a sygus universal variable */
+class CVC4_PUBLIC DeclareSygusVarCommand : public DeclarationDefinitionCommand
+{
+ protected:
```
Comment (ajreynol) [inline src/smt/command.h]: Let's make these new command conform to the style guidelines, where protected goes below public.

Same goes for the others.

File: src/smt/smt_engine.h [hunk 10]
```diff
@@ -608,11 +608,50 @@ class CVC4_PUBLIC SmtEngine {
...
+   * place holders :
+   *
+   * PRE(x) -> INV(x)
+   * INV() ^ TRANS(x, x') -> INV(x')
+   * INV(x) -> POST(x)
+   *
+   * The regular and primed variables are retrieved from the declaration of the
+   * invariant-to-synthesize.
+   */
+  void assertSygusInvConstraint(const std::vector<Expr>& place_holders);
```
Comment (ajreynol) [inline src/smt/smt_engine.h]: Maybe name `place_holders` -> `predicates`?

Comment (HanielB) [reply src/smt/smt_engine.h]: That's a good idea, thanks. I was not happy with this name.

File: src/smt/smt_engine.cpp [hunk 11]
```diff
@@ -497,6 +497,31 @@ class SmtEnginePrivate : public NodeManagerListener {
...
+  std::vector<Type> d_sygusPrimedVarTypes;
+  /** sygus constraints */
+  std::vector<Node> d_sygusConstraints;
+  /** functions-to-synthesize */
+  std::vector<Node> d_sygusFunSymbols;
+  /** maps functions-to-synthesize to their respective input variables lists */
+  std::map<Node, std::vector<Node>> d_sygusFunVars;
+  /** maps functions-to-synthesize to their respective syntactic restrictions
+   *
+   * If function has syntactic restrictinos, these are encoded as a SyGuS
```
Comment (ajreynol) [inline src/smt/smt_engine.cpp]: restrictions

File: src/smt/smt_engine.h [hunk 12]
```diff
@@ -608,11 +608,50 @@ class CVC4_PUBLIC SmtEngine {
...
+   * place holder predicates :
+   *
+   * PRE(x) -> INV(x)
+   * INV() ^ TRANS(x, x') -> INV(x')
+   * INV(x) -> POST(x)
+   *
+   * The regular and primed variables are retrieved from the declaration of the
+   * invariant-to-synthesize.
+   */
+  void assertSygusInvConstraint(const std::vector<Expr>& predicates);
```
Comment (ajreynol) [inline src/smt/smt_engine.h]: Actually, maybe we can have an explicit argument list here, pre, trans, post, inv?

File: src/smt/command.h [hunk 13]
```diff
@@ -613,29 +613,251 @@ class CVC4_PUBLIC QueryCommand : public Command
...
+ * invariant-to-synthesize, precondition, transition relation and condition.
+ *
+ * The actual constraint must be built such that the invariant is not stronger
+ * than the precondition, not weaker than the postcondition and inductive
+ * w.r.t. the transition relation.
+ */
+class CVC4_PUBLIC SygusInvConstraintCommand : public Command
+{
+ public:
+  SygusInvConstraintCommand(const std::vector<Expr>& predicates);
```
Comment (ajreynol) [inline src/smt/command.h]: Let's make this pre, trans, post, inv.

File: src/smt/smt_engine.h [hunk 14]
```diff
@@ -608,11 +608,53 @@ class CVC4_PUBLIC SmtEngine {
    */
   std::vector<Expr> getUnsatAssumptions(void);
 
+  /*------------------- sygus commands  ------------------*/
+
+  /** adds a variable declaration */
```
Comment (ajreynol) [inline src/smt/smt_engine.h]: Improve this comment.

File: src/smt/smt_engine.h [hunk 15]
```diff
@@ -608,11 +608,53 @@ class CVC4_PUBLIC SmtEngine {
...
+  void declareSygusVar(const std::string& id, Expr var, Type type);
+  /** stores information for debugging sygus invariants setup
+   *
+   * Since in SyGuS the commands "declare-primed-var" are not necessary for
+   * building invariant constraints, we only use them to check that the number
+   * of variables declared corresponds to the number of arguments of the
+   * invariant-to-synthesize.
+   */
+  void declareSygusPrimedVar(const std::string& id, Type type);
+  /** adds a function variable declaration */
```
Comment (ajreynol) [inline src/smt/smt_engine.h]: Improve this comment.

File: src/smt/smt_engine.h [hunk 16]
```diff
@@ -608,11 +608,53 @@ class CVC4_PUBLIC SmtEngine {
...
+  void declareSygusFunctionVar(const std::string& id, Expr var, Type type);
+  /** adds a function-to-synthesize declaration */
```
Comment (ajreynol) [inline src/smt/smt_engine.h]: Improve this, explain the arguments.

**Pull request merged**